feat: ✨ catniss and her friends are able to vote

gchamb opened the pull request.

gchamb committed:
feat(firestore): ✨ allow user to vote for others and not themselves or the same person

gchamb committed:
feat(client): ✨ create vote component, collection, and add vote function

gchamb committed:
feat(functions): ✨ add vote functionality on backend

- created vote collection
- listen to voting with trigger function
- created findVoteoff function to find the results of the voting

gchamb committed:
feat(functions): ✨ add timers for voting and result

gchamb committed:
feat: ✨ create result component to display the voted of player

- create result component
- create verify expiration function for result

gchamb committed:
feat(functions): ✨ type changes

- changed app.ts to ignore undefined. It made it easier for me to simply put winner as property since it's either undefined | Role.

gchamb committed:
style: 🎨 prettier check

gchamb committed:
fix(functions): 🐛 put appropriate type and comment

nstringham commented on functions/src/firestore-types/lobby.ts:
```diff
@@ -69,6 +74,14 @@ export type Lobby = {
    * expiration time of the current phase with a timer
    */
   expiration?: Timestamp;
+  /**
+   * the player that was voted off for the round
+   */
+  votedOff?: string;
+  /**
+   * the role that won the game
+   */
+  winner?: Role;
```
cam we have blank lines between properies please

nstringham commented on functions/src/firestore-types/lobby.ts:
```diff
@@ -69,6 +74,14 @@ export type Lobby = {
    * expiration time of the current phase with a timer
    */
   expiration?: Timestamp;
+  /**
+   * the player that was voted off for the round
+   */
+  votedOff?: string;
```
there should be a note here about the `NONE` specail value

maybe even set the type to `string | "NONE"`

nstringham replied:
this should be a uid

nstringham commented on functions/src/firestore-types/lobby.ts:
```diff
@@ -40,7 +44,8 @@ export const GAME_STATE_DURATIONS: { [state in GameState]: number } = {
   PROMPT: 60,
   CHAT: 2 * 60,
-  VOTE: 3 * 60,
+  VOTE: 30, //TODO: replace with correct timer when getting merged
```
don't forget this

nstringham commented on client/src/lib/firebase/vote.ts:
```diff
@@ -0,0 +1,8 @@
+import { getVoteCollection } from "./firestore-collections";
+
+export function addVote(lobbyCode: string, userId: string, target: string) {
+  setDoc(doc(getVoteCollection(lobbyCode), userId), {
```
should await or return so that erros can be caught and displayed to the uesr

nstringham commented on functions/src/lobby.ts:
```diff
@@ -193,7 +195,19 @@ export const onLobbyUpdate = functions.firestore.document("/lobbies/{code}").onU
-      firestore.Timestamp.now().toMillis() + GAME_STATE_DURATIONS.CHAT * 1000
+      firestore.Timestamp.now().toMillis() + GAME_STATE_DURATIONS.END * 1000 // TODO: replace with correct timer when doing pull request
+    );
+    lobbyDocRef.set({ expiration }, { merge: true });
```
should probobly use update instead of set with merge: true

nstringham commented on functions/src/vote.ts:
```diff
@@ -0,0 +1,51 @@
+
+  // find the player with the most votes
+  const [most, secondMost]: Player[] = JSON.parse(JSON.stringify(players)).sort((a: Player, b: Player) => {
+    if (a.votes == undefined) {
+      a.votes = 0;
+    }
+    if (b.votes == undefined) {
+      b.votes = 0;
+    }
+
+    return b.votes - a.votes;
```
you could just do
```ts
return b.votes ?? 0 - a.votes ?? 0;
```

nstringham commented on functions/src/vote.ts:
```diff
@@ -0,0 +1,51 @@
+        player.avatar == most.avatar &&
+        player.displayName == most.displayName &&
+        player.promptAnswer == most.promptAnswer &&
+        player.votes == most.votes
```
technically only need to check votes because it is unique

nstringham commented on functions/src/vote.ts:
```diff
@@ -0,0 +1,51 @@
+      players[mostIndex] = most;
+    }
+    // save their name
+    votedOff = most.displayName;
```
this should be a uid

nstringham commented on functions/src/vote.ts:
```diff
@@ -0,0 +1,51 @@
+    // replace the shallow one copy object with the deep copy object changes
+    if (mostIndex != undefined) {
+      players[mostIndex] = most;
+    }
```
why not just do `players[mostIndex].alive = false`

nstringham commented on functions/src/vote.ts:
```diff
@@ -0,0 +1,51 @@
+  transaction.update(lobbyDocRef, {
+    state: "RESULT",
+    players: players,
+    votedOff: votedOff == false ? "NONE" : votedOff,
```
if you initalized `votedOff`to `undifnined` then you can just do `votedOff ?? "NONE"`

nstringham commented on functions/src/vote.ts:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  // update the state of the game
+  transaction.update(lobbyDocRef, {
```
should return or await this

nstringham commented on functions/src/lobby.ts:
```diff
@@ -341,6 +399,12 @@ export const verifyExpiration = functions.https.onCall(async (data, context) =>
       await deleteChatRooms(lobby, lobbyDocRef, transaction);
     }
+    if (lobby.state === "VOTE") {
+      findVoteOff(lobby, lobbyDocRef, transaction);
```
should await this

gchamb replied:
it returns void tho not Promise<void>

nstringham commented on functions/src/lobby.ts:
```diff
@@ -305,6 +319,50 @@ export const collectPromptAnswers = functions.firestore
+  // check if the after change exists
+  // we want to do this because onwrite is for oncreation, onupdate, and ondelete
+  if (!change.after.exists) {
+    return {};
```
what is the `{}` for?

nstringham commented on functions/src/lobby.ts:
```diff
@@ -305,6 +319,50 @@ export const collectPromptAnswers = functions.firestore
+    }
+
+    // increment new target
+    if (target.votes == undefined) {
+      target.votes = 1;
+    } else {
+      target.votes += 1;
+    }
```
would it be better to make `vote` not optional and just set it to `0` from the very beginging?

gchamb replied:
it would be better imo

nstringham commented on client/src/components/Vote.svelte:
```diff
@@ -0,0 +1,91 @@
+    {#each lobby.players as { avatar, displayName, votes }, i}
+      <div class="vote-container">
+        <button class="avatar" on:click={() => addVote(lobbyCode, $user?.uid ?? "", lobby.uids[i])}>
+          <img src="/avatars/{avatar}.webp" alt="cat picture {avatar}" />
```
please use the avatar alt text from `avatar.ts`

nstringham commented on client/src/components/Vote.svelte:
```diff
@@ -0,0 +1,91 @@
+  <div class="voting-grid">
+    {#each lobby.players as { avatar, displayName, votes }, i}
+      <div class="vote-container">
+        <button class="avatar" on:click={() => addVote(lobbyCode, $user?.uid ?? "", lobby.uids[i])}>
```
should be disabled if it is the currently logged in player

gchamb committed:
refactor(functions): ♻️ voted from optional to required and made changes to reflect that

- added space in between properties
- switched from set to update from the timers
- added awaits

gchamb committed:
refactor(client): ♻️ use avatar alt text and change the result component to reflect votedoff as uid

gchamb committed:
refactor(functions): ♻️ subtle changes to reflect votes being required

gchamb committed:
refactor(functions): ♻️ revert game state for chat back

gchamb committed:
Merge branch 'development' into gchamb/voting

gchamb committed:
refactor(functions): 🔥 remove colliding parts

gchamb committed:
refactor(functions): ♻️ add space

gchamb committed:
feat(functions): 🎨 added note for voted off

nstringham commented on functions/src/lobby.ts:
```diff
@@ -195,13 +197,21 @@ export const onLobbyUpdate = functions.firestore.document("/lobbies/{code}").onU
   if (lobby.state == "CHAT" && oldLobby.state != "CHAT") {
     const expiration = firestore.Timestamp.fromMillis(
-      firestore.Timestamp.now().toMillis() + GAME_STATE_DURATIONS.CHAT * 1000
+      firestore.Timestamp.now().toMillis() + GAME_STATE_DURATIONS.END * 1000 // TODO: replace with correct timer when doing pull request
```
don't forget this

nstringham commented on functions/src/vote.ts:
```diff
@@ -0,0 +1,36 @@
+  }
+
+  // update the state of the game
+  await transaction.update(lobbyDocRef, {
```
don't need this await

nstringham commented on functions/src/lobby.ts:
```diff
@@ -311,6 +321,45 @@ export const collectPromptAnswers = functions.firestore
+  });
+});
+
 export const verifyExpiration = functions.https.onCall(async (data, context) => {
```
this should also have a return type

gchamb committed:
refactor(functions): ♻️ changed chat timer back

gchamb committed:
refactor(functions): ♻️ remove async and await
